### PR TITLE
Add initial support for differentiable optimizers

### PR DESCRIPTION
**Pull request by emcastillo**
Adds the `differentiable` argument, a method for updating parameters in an existing optimizer, and a template for testing the differentiability of multiple optimizers.

This is all based in discussions with @albanD & @jbschlosser

### CODE REVIEW

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -100,7 +101,8 @@ def __init__(self, params, lr=required, momentum=0, dampening=0,
 
         defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
                         weight_decay=weight_decay, nesterov=nesterov,
-                        maximize=maximize, foreach=foreach)
+                        maximize=maximize, foreach=foreach,
+                        differentiable=differentiable)
```
Comment (jbschlosser) [inline torch/optim/sgd.py:105]: Do we need `group.setdefault('differentiable', False)` in `__setstate__` below as well?

File: torch/optim/sgd.py [hunk 2]
```diff
@@ -195,16 +201,17 @@ def sgd(params: List[Tensor],
...
-         d_p_list,
-         momentum_buffer_list,
-         weight_decay=weight_decay,
-         momentum=momentum,
-         lr=lr,
-         dampening=dampening,
-         nesterov=nesterov,
-         has_sparse_grad=has_sparse_grad,
-         maximize=maximize)
+    func(
```
Comment (jbschlosser) [inline torch/optim/sgd.py:204]: Was this a lint change? :p

File: torch/optim/optimizer.py [hunk 3]
```diff
@@ -293,7 +284,7 @@ def add_param_group(self, param_group):
             if not isinstance(param, torch.Tensor):
                 raise TypeError("optimizer can only optimize Tensors, "
                                 "but one of the params is " + torch.typename(param))
-            if not param.is_leaf:
+            if not self.defaults.get('differentiable', False) and not param.is_leaf:
```
Comment (jbschlosser) [inline torch/optim/optimizer.py:287]: Do we want to remove the `is_leaf` restriction entirely or is there still some reason to keep it around? cc @albanD

File: torch/optim/optimizer.py [hunk 4]
```diff
@@ -317,3 +308,20 @@ def add_param_group(self, param_group):
             raise ValueError("some parameters appear in more than one parameter group")
 
         self.param_groups.append(param_group)
+
+    def update_parameters(self, params):
```
Comment (jbschlosser) [inline torch/optim/optimizer.py:312]: Hm, I think we have to be very careful introducing this new public API.
* I think the name should be tweaked since applying gradient updates could be thought of as "updating parameters" and that's not what's being done here (maybe `swap_parameters()` or similar is better?).
* This mutating API allows the "parameters remain constant throughout the optimizer lifetime" invariant that used to hold for optimizers to be broken. I wonder if there are areas in the code that assume this invariant and break if it doesn't hold (e.g. LR schedulers)?

For stateful optimizers such as SGD, is this the only way to support optimizing on a different set of params?

File: test/test_optim.py [hunk 5]
```diff
@@ -2665,5 +2665,26 @@ def test_bn_update_eval_momentum(self):
 
 instantiate_parametrized_tests(TestLRScheduler)
 
+def _diff_fn(p, grad, opt_differentiable_state, opt):
+    p = p.clone()
+    p.grad = grad
+    opt.update_parameters([p])
+    opt.state.update(opt_differentiable_state)
```
Comment (jbschlosser) [inline test/test_optim.py:2672]: Is this necessary or is the state mutated during the `step()` call?

**Review by jbschlosser — COMMENTED** [review]
Thanks for working on this! I think this is mostly in-line with what we discussed: manual cloning required by the user + an API for mutating stateful optimizers to run over a different set of params. Added a couple concerns about the new API below, but nothing that can't fundamentally be addressed.

File: torch/optim/optimizer.py (hunk 3 above)
Comment (albanD) [reply torch/optim/optimizer.py:287]: The original reason is still there: you should not give a Tensor to your optimizer that will not get a .grad field.
I think a good way to relax this would be to do `not (param.is_leaf or param.retains_grad)` for the second part of the condition. And disabling it for differentiable sounds good for sure.

File: torch/optim/optimizer.py (hunk 4 above)
Comment (albanD) [reply torch/optim/optimizer.py:312]: Also this fully resets the current state of the optimizer.
For the purpose of this PR and the meta learning example we have, I don't think we need such a function.
Can you simply create the optimizer inside the test function?

File: test/test_optim.py [hunk 6]
```diff
@@ -2665,5 +2665,26 @@ def test_bn_update_eval_momentum(self):
 
 instantiate_parametrized_tests(TestLRScheduler)
 
+def _diff_fn(p, grad, opt_differentiable_state, opt):
+    p = p.clone()
```
Comment (albanD) [inline test/test_optim.py:2669]: My mistake from the sample I shared, but I think you want to clone the opt_differentiable_state as well?

File: test/test_optim.py [hunk 7]
```diff
@@ -2665,5 +2665,26 @@ def test_bn_update_eval_momentum(self):
...
+    p.grad = grad
+    opt.update_parameters([p])
+    opt.state.update(opt_differentiable_state)
+    opt.step()
+    return p
```
Comment (albanD) [inline test/test_optim.py:2674]: You can return the opt_differentiable_state here as well to make sure all gradients flow properly

File: test/test_optim.py [hunk 8]
```diff
@@ -2665,5 +2665,26 @@ def test_bn_update_eval_momentum(self):
...
+
+
+class TestDifferentiableOptimizer(TestCase):
+
+    def test_sgd(self):
+        p = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        grad = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        mbuff = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        state = {'momentum_buffer': mbuff}
+        opt = torch.optim.SGD([p], lr=0.9, differentiable=True)
```
Comment (albanD) [inline test/test_optim.py:2684]: moving this inside _diff_fn should remove the need for the param swapping function.

File: torch/optim/sgd.py (hunk 1 above)
Comment (albanD) [reply torch/optim/sgd.py:105]: What about we add this field directly to the base `Optimizer` class (in the `__setstate__` in particular).
We want this to be there in the end anyways and that will avoid any BC issue where we forget to add it for a specific optimizer.

File: torch/optim/sgd.py [hunk 9]
```diff
@@ -161,6 +165,8 @@ def step(self, closure=None):
                 state = self.state[p]
                 state['momentum_buffer'] = momentum_buffer
 
+        if not self.defaults['differentiable']:
```
Comment (albanD) [inline torch/optim/sgd.py:168]: nit: it would be better to have this in a try: finally: bloc
Maybe you can have a small context manager specialized for this that you apply to the step function and that we can re-use for all optimizers?

File: torch/optim/sgd.py (hunk 1 above)
Comment (emcastillo) [reply torch/optim/sgd.py:105]: The `__setstate__` is only invoked when `load_state_dict` is called. How to add a `self._differentiable` to the `Optimizer.__init__`?

File: torch/optim/optimizer.py (hunk 4 above)
Comment (emcastillo) [reply torch/optim/optimizer.py:312]: > For stateful optimizers such as SGD, is this the only way to support optimizing on a different set of params?

I don't think of another possible way to support this other than passing a `param_groups` kwarg to `step` to act in a functional way

File: test/test_optim.py (hunk 5 above)
Comment (emcastillo) [reply test/test_optim.py:2672]: step call mutates inplace some of the values such as the optimizer momentum buffers, So we need to replace these somehow.

File: torch/optim/optimizer.py (hunk 4 above)
Comment (emcastillo) [reply torch/optim/optimizer.py:312]: > Can you simply create the optimizer inside the test function?

I think this is fine, but do we want to create an optimizer object for every step?
Since the user will be able to provide all the needed internal `state` values creating two optimizers for subsequent updates should be equivalent to calling `.step` on the same optimizer twice. However, the way that LR schedulers work could be affected?

File: test/test_optim.py (hunk 8 above)
Comment (emcastillo) [reply test/test_optim.py]: done!

File: test/test_optim.py (hunk 6 above)
Comment (emcastillo) [reply test/test_optim.py:2669]: done!

File: test/test_optim.py (hunk 7 above)
Comment (emcastillo) [reply test/test_optim.py]: done!

File: torch/optim/sgd.py (hunk 9 above)
Comment (emcastillo) [reply torch/optim/sgd.py]: done!

File: torch/optim/sgd.py (hunk 2 above)
Comment (emcastillo) [reply torch/optim/sgd.py]: fixed :D

File: test/test_optim.py [hunk 10]
```diff
@@ -2680,5 +2680,25 @@ def test_bn_update_eval_momentum(self):
...
+
+
+class TestDifferentiableOptimizer(TestCase):
+
+    def test_sgd(self):
+        p = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        grad = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        mbuff = torch.rand(10, requires_grad=True, dtype=torch.float64)
+        state = {'momentum_buffer': mbuff}
+        gradcheck(_diff_fn, (p, grad, state, torch.optim.SGD, {'lr': 0.9, 'differentiable': True}))
```
Comment (albanD) [inline test/test_optim.py:2700]: gradcheck will not "unpack" data structures so the content of `state` will not be checked here.
It is fine with the inputs being captured though, so the simplest change you can do here is add a new argument to `_diff_fn` that is `*ignored`. And pass in `, *state.values())` as part of the arguments.

**Review by albanD — COMMENTED** [review]
Sounds good to me. Only a small update on the test!

File: test/test_optim.py (hunk 10 above)
Comment (albanD) [reply test/test_optim.py]: Yep exactly!

**Review by albanD — APPROVED** [review]
SGTM!
It would be nice to have a small comment explaining why we have this ignored arg in the test :D

**Comment by emcastillo** [comment]
Added the comment!
